lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -2041,15 +2041,14 @@ SILFunction *GenericFuncSpecializer::lookupSpecialization() {
                                   SILLinkage::Shared);
   }
   if (SpecializedF) {
-    if (ReInfo.getSpecializedType() != SpecializedF->getLoweredFunctionType()) {
+    auto expectedTy = ReInfo.getSpecializedType()->getCanonicalType();
+    auto foundTy = SpecializedF->getLoweredFunctionType()->getCanonicalType();
+    if (expectedTy != foundTy) {
       llvm::dbgs() << "Looking for a function: " << ClonedName << "\n"
-                   << "Expected type: " << ReInfo.getSpecializedType() << "\n"
-                   << "Found    type: "
-                   << SpecializedF->getLoweredFunctionType() << "\n";
+                   << "Expected type: " << expectedTy << "\n"
+                   << "Found    type: " << foundTy << "\n";
+      llvm_unreachable("Specialized function does not match expected type.");
```

Comment:
I'm not sure if llvm_unreachable is the right thing to do here. In NDEBUG builds it might get optimized away.

```
/// Marks that the current location is not supposed to be reachable.
/// In !NDEBUG builds, prints the message and location info to stderr.
/// In NDEBUG builds, if the platform does not support a builtin unreachable
/// then we call an internal LLVM runtime function. Otherwise the behavior is
/// controlled by the CMake flag
///   -DLLVM_UNREACHABLE_OPTIMIZE
/// * When "ON" (default) llvm_unreachable() becomes an optimizer hint
///   that the current location is not supposed to be reachable: the hint
///   turns such code path into undefined behavior.  On compilers that don't
///   support such hints, prints a reduced message instead and aborts the
///   program.
/// * When "OFF", a builtin_trap is emitted instead of an
//    optimizer hint or printing a reduced message.
```

I think it's safer to use `ASSERT`